lib/rubocop/cop/lint/reduce_accumulator.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Looks for `reduce` or `inject` blocks where the value returned (implicitly or
+      # explicitly) does not include the accumulator. A block is considered valid as
+      # long as at least one return value includes the accumulator.
+      #
+      # If the accumulator is not included in the return value, then the entire
+      # block will just return a transformation of the last element value, and
+      # could be rewritten as such without a loop.
+      #
+      # Also catches instances where an index of the accumulator is returned, as
+      # this may change the type of object being retained.
+      #
+      # NOTE: For the purpose of reducing false positives, this cop only flags
+      # returns in `reduce` blocks where the element is the only variable in
+      # the expression (since we will not be able to tell what other variables
+      # relate to via static analysis).
+      #
+      # @example
+      #
+      #   # bad
+      #   (1..4).reduce(0) do |acc, el|
+      #     el * 2
+      #   end
+      #
+      #   # bad, may raise a NoMethodError after the first iteration
+      #   %w(a b c).reduce({}) do |acc, letter|
+      #     acc[letter] = true
+      #   end
```

Comment:
I concede that that is a valid use that will be registered by the cop, but I'd argue that this is going to be *much* less prevalent than mistakenly forgetting to return the accumulator when setting a hash key.

I see a few options:
1. We can remove this behaviour (I think this isn't ideal, because I do think this is still a good check)
2. We can put it behind a conf variable that is not enabled by default
3. We can accept the false positives (if this was my code I'd just `# rubocop:disable` it, but maybe that's suboptimal).

What do we think? /cc @bbatsov

Reply:
> forgetting to return the accumulator when setting a hash key.

BTW this is a double error, as `inject` probably shouldn't be used in the first place in that case (see `Style/EachWithObject`)

Reply:
Definitely true, but in this case `Style/EachWithObject` doesn't register (because the accumulator is not being returned).